src/backend/catalog/storage_pending_deletes_redo.c

```diff
@@ -0,0 +1,333 @@
+ */
+void
+PdlRedoXLogRecord(XLogRecord *record)
+{
+	Assert(record);
+
+	if (IsBootstrapProcessingMode() || !gp_track_pending_delete)
+		return;
+
+	PendingRelXactDeleteArray *arr = (PendingRelXactDeleteArray *) XLogRecGetData(record);
+
+	TransactionId oldest_xid = ShmemVariableCache->oldestXid;
+
+	Assert(arr->count);
+
+	for (int i = 0; i < arr->count; i++)
+	{
+		PendingRelXactDelete *pd = (PendingRelXactDelete *) &(arr->array[i]);
+
+		/*
+		 * This function should check transaction status before adding
+		 * relfilenode to a pendingDeletesRedo hash table. Concurrent xlog
+		 * inserts (concurrent to a checkpointing process) of commit or abort
+		 * xlog records may out out-date pending deletes list. We don't want
+		 * to use aggressive locking of shared structures in order to avoid
+		 * performance drawbacks of concurrent commits or aborts. So the
+		 * strategy is to double-check relfilenodes with it's transaction
+		 * status. If it's TRANSACTION_STATUS_IN_PROGRESS, then it's
+		 * permitted to delete files (it's orphaned), if it's in some other
+		 * status - don't touch it. Also we should check transaction xid
+		 * doesn't cross "freeze horizon" and compare it with current
+		 * oldestXid value. Motivation of this check is that clog might get
+		 * truncated after REDO point and before replaying XLOG_PENDING_DELETE
+		 * record (though that looks like unlikely will happen in real-world,
+		 * but still needs to be considered as possible scenario). So in that
+		 * case we can't rely on xid status of that frozen transactions.
+		 * Second point is that there is no way that clog would be truncated
+		 * when transaction is in progress, so it's either been committed or
+		 * aborted before that.
+		 */
+
+		if (TransactionIdPrecedes(pd->xid, oldest_xid))
+			elog(WARNING, "Prevented adding node for XLOG_PENDING_DELETE record for xid: %u, oldestXid: %u",
+				 pd->xid, oldest_xid);
+		else
+		{
+			XLogRecPtr	result;
+			XidStatus	status = TransactionIdGetStatus(pd->xid, &result);
+
+			if (status == TRANSACTION_STATUS_IN_PROGRESS)
+				PdlRedoAdd(pd);
+			else
+				elog(WARNING, "Prevented adding node for XLOG_PENDING_DELETE record for xid: %u, status: %d",
+					 pd->xid, status);
+		}
+	}
+}
+
+static void
+PdlRedoRemove(TransactionId xid)
+{
+	if (IsBootstrapProcessingMode() ||
+		(xid == InvalidTransactionId) ||
+		(NULL == pendingDeletesRedo) ||
+		!gp_track_pending_delete)
+		return;
+
+	PendingDeleteHtabNode *entry = (PendingDeleteHtabNode *)
+		hash_search(pendingDeletesRedo, &xid, HASH_REMOVE, NULL);
+
+	if (entry)
+		list_free_deep(entry->relnode_list);
+}
+
+/*
+ * This function removes pending delete nodes from redo hash-table
+ * (pendingDeleteRedo) for a given transaction identified by it's xid and
+ * sub-transactions (if there are). It is called from xact_redo_commit_internal,
+ * xact_redo_distributed_commit or xact_redo_abort.
+ * (TODO: and RemovePendingDeletesForPreparedTransactions)
```

Comment:
It is redundant to write in a comment where a function is called from, because we can see this places in the code. Comment should contain unobvious information only